RFD 107 - Desktop Access Performance (#22073)
gravitational/teleport: rfd/0107-desktop-access-performance-research.md

### Remote Desktop Protocol: Graphics Pipeline Extension  [MS-RDPEGFX](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegfx/da5c75f9-cd99-450c-98c4-014a496942b0)

The [MS-RDPEGFX] extension introduces a new way of drawing data by creating surfaces and using commands that utilize these surfaces to draw frames rather than just rendering bitmap data on the screen. This extension efficiently encode and transmit graphic display data from the server to the client. It alsow allows caching data locally to improve performance, can use different, newer, and more performant codecs to encode data e.g:
- RemoteFX
- ClearCodec
- MPEG-4 AVC/H.264

In essence, this protocol works in this way:
- create surfaces
- associate surfaces to the portions of graphics output buffer (this is what an user sees) that can be of different sizes
- decode graphic data for surface and render it on the graphics output buffer

> There is already one implementation of the extension in the Rust language: [ironrdp-gui-client](https://github.com/Devolutions/IronRDP/tree/master/ironrdp-gui-client), but it is using native gui for rendering by utilising opengl. 

Reviewer: What makes you think this is using `MS-RDPEGFX`? From what I see, it's using the same `process_connection_sequence` and therefore the same capabilities exchange as other clients in the repo, ergo it's using the same version of the protocol.

That said, [apparently some of GFX was implemented for video streaming](https://github.com/Devolutions/IronRDP/issues/76#issuecomment-1433676943).

Author: Yes that's right that it's using the `process_connection_sequence` but you can pass your own capabilites or [enable the avc444 / h264 in cofing](https://github.com/Devolutions/IronRDP/blob/master/ironrdp-gui-client/src/config.rs#L120) and then that config is passed to the active session processor.

and then:
- [start gui](https://github.com/Devolutions/IronRDP/blob/master/ironrdp-gui-client/src/main.rs#L153)
- [create renderer and pass window to it](https://github.com/Devolutions/IronRDP/blob/master/ironrdp-gui-client/src/gui.rs#L82)
- and then this render starts two threads, one for [processing [MS-RDPGFX] messages](https://github.com/Devolutions/IronRDP/blob/master/ironrdp-renderer/src/renderer.rs#L147) and [another for drawing](https://github.com/Devolutions/IronRDP/blob/master/ironrdp-renderer/src/renderer.rs#L152)

Anyway if you don't enable avc444 / h264 it'll use the bitmaps encoded in the RemoteFX. Another thing is that this gui client doesn't work on macOS without a lot of fixes

Reviewer: I see, thanks for linking me to it

               |                                        |        |                   |
+----------------------+     +------------------+  +------------------+     +------------------+
|                      |     |                  |  | (rdp-rs & b-2-p) |     |                  |
|  (TDP client)        |     |                  |  |    Teleport      |     |                  |
|  User's Web Browser  ------|  Teleport Proxy -----  Windows Desktop ------|  Windows Desktop |
|                      |     |                  |  |     Service      |     |                  |
+----------------------+     +------------------+  +------------------+     +------------------+
```

Meanwhile, IronRDP's default setup looks like this when built for a native target (MacOS/linux/Windows):

```
                                RDP
               ------------------------------------------
               |                                        |
+----------------------+                           +------------------+
|                      |                           |                  |
|  (IronRDP)           |                           |                  |

> |  User's Web Browser  -----------------------------  Windows Desktop |

Reviewer 2: ```suggestion
|  User's Workstation  -----------------------------  Windows Desktop |
```

---
authors: Łukasz Kozłowski (lukasz.kozlowski@goteleport.com)
state: draft
---

# RFD 107 - Desktop Access Performance

## Required Approvers
Engineering: @zmb3 && @ibeckermayer

## What

This RFD explores possible solutions to the current performance problems in the destkop
access module.

## Why

We want to improve user experiance when using the desktop access module. Currently the main issue
is insufficent performance of the video rendering which results in choppier experience or noticeable stuttering. Each frame in the RDP protocol consists of (in most cases) 64x64 pixel bitmaps that are assembeld into the full frame. All of these bitmaps needs to be send by the server, received and processed by the client. There are different ways in which we can reduce time to render a frame.
One of the ways is to reduce number of bitmaps that we need to send over the wire. The other way is to reduce the time to process individual bitmap. In this document we'll discuss some solutions that might fix these problems.

### RDP Bitmap caching

> One of the ways to improve the performance of Teleport Desktop Access is to reduce the amount of data sent between the proxy and the web browser and to reduce the number of messages with bitmap data that needs processing. One RDP protocol extensions address the problem of reducing latency and bandwidth usage. It's called [[MS-RDPEGDI] Remote Desktop Protocol: Graphics Device Interface (GDI) Acceleration Extensions](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/745f2eee-d110-464c-8aca-06fc1814f6ad)

Reviewer 2: @ibeckermayer also pointed out [MS-RDPEGFX](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegfx/da5c75f9-cd99-450c-98c4-014a496942b0)

How does MS-RDPEGFX compare to MS_RDPEGDI?

Author: First of all, these two extensions are unrelated to each other. 

The "Remote Desktop Protocol: Graphics Device Interface (GDI)
Acceleration Extensions [MS-RDPEGDI]" was released on 2/22/2007 and is available starting from the RDP 6.0.

The  "Remote Desktop Protocol: Graphics Pipeline Extension [MS-RDPEGFX]" was released on 12/16/2011 and is available starting from the RDP 8.0.

Both extensions improve performance and reduce the number of operations the client needs to do to render a full frame. They define and use different primitives to achieve this goal.

The [MS-RDPEGDI] extension specifies drawing operations, like combining bitmap patterns using raster operations and painting glyphs, rectangles, lines, and ellipses. It also allows transferring portions of the screen area from one point to another. It also defines bitmap caches and operations to store and load bitmap data from these caches.

On the other hand, the [MS-RDPEGFX] extension introduces surfaces and operations that act upon them. There is also a graphics output buffer which is an end-user visible bitmap. 
It works more or less as follows: 
- create surfaces on the client
- transfer encoded bitmap data from the server to a client-side destination surface
- map surfaces to the graphics output buffer
- render frame

There are also operations that allow copying bitmap data from one surface to another. This extension also defines bitmap caches and operations that can store and load bitmap data in these caches. It also uses different and newer codecs to encode bitmap information e.g:
- RemoteFX
- ClearCodec
- MPEG-4 AVC/H.264

The client using this extension might be more performant, but it defines a new way of rendring bitmap data. It also requires support of a lot more operations and a newer version of the RDP compared to only using caches operations from the [MS-RDPEGDI].

Reviewer: > It also uses different and newer codecs to encode bitmap information e.g:
>
> - RemoteFX
> - ClearCodec
> - MPEG-4 AVC/H.264

Technically RemoteFX is a part of [MS-RDPEGDI]: see [codecGUID](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/86507fed-a0ee-4242-b802-237534a8f65e)

This slide is helpful
<img width="1247" alt="image" src="https://user-images.githubusercontent.com/13578537/221306753-24971be0-3f3e-48b3-9bc8-e94360f63ae0.png">

([source](https://interopevents.blob.core.windows.net/uploads/PDFs/2019/RDPIO/RDP%20IO%20Lab%20-%20RDP%20Update%202019%20-Denis%20Gundarev.pdf))

This extension reduces bandwidth usage by encoding the drawing operations that produce an image instead of sending the actual bitmap data. Multiple drawing operations are introduced in this extension, but this document focuses on only two operations that interact with the bitmap caches. The server and the client uses bitmap caches to store graphic bitmaps.

The way it works is both the server and the client has a bitmap cache, but it's the server responsibility to control the contents of both bitmap caches. The server can instruct the client to store bitmap data inside a bitmap cache. It can also tell the client to load stored data from the cache and render in on the screen. The main benefit of this architecture is that the stored bitmap data in the bitmap cache can be loaded and rendered multiple times without the need to send data over again. Only the render bitmap operation from the cache needs have to be transmited over the wire. On top of that single message from the RDP server can contain tens or even hundreds of drawing operations. It can significantly reduce the latency and bandwidth usage.

> There are two types of caches specified in the extension. The first one, called `Revision 1` only support memory-based caching and defaults to 6MB of cache storage for 32bpp bitmaps. The other one, called `Revision 2` support persistent disk caching in addition to memory caching and defaults to 40MB of cache storage for 32bpp bitmaps.

Author: I thought that might use 32bpp bitmaps. Anyway, it's 20MB of cache for 16bps bitmaps. Here's a full breakdown of caches and their [limits](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/2bf92588-42bd-4527-8b3e-b90c56e292d2)

Our choice for a support of caching is to support `Revision 2` cache since it allows to store more data inside its cache.

#### How it works on the protocol level?

> Two new types of messages needs to be handled by our library in order to enable support for bitmap caching. Message which stores bitmap data inside a bitmap cache, called `Cache Bitmap - Revision 2` and message which renders stored bitmap data in the bitmap cache, called `MemBlt`.
> Message which stores bitmap data inside a bitmap cache contains bitmap data and cache index generated by the server. On the other hand, operation that renders the bitmap only has cache index and position on the screen where to render the bitmap.
> The flow of the messages is controlled by the server so it is not possible to have a message which renders a bitmap that uses cache key that hasn't been yet used to store a bitmap. 

Reviewer: ```suggestion
Two new types of messages need to be handled by our RDP library in order to enable support for bitmap caching. One that stores bitmap data inside a bitmap cache, called `Cache Bitmap - Revision 2`, and another that renders a previously stored bitmap, called `MemBlt`.
`Cache Bitmap - Revision 2` contains the bitmap's data and a cache index generated by the server. On the other hand, `MemBlt` contains only the cache index and the position on the screen where the bitmap should be rendered.
```

That last sentence seems unnecessary to point out.

#### What changes are required to support caching?

> Teleport Desktop Access protocol, TDP, needs to be extended with two messages: 

Reviewer: ```suggestion
Teleport Desktop Protocol (TDP) needs to be extended with two messages: 
```

> #### 29 - bitmap cache store

Reviewer: ```suggestion
##### 29 - bitmap cache store
```

> #### 29 - bitmap cache store

Author: Yes, you're right. I'll change it so we are clear that we're using PNGs at the TDP layer.

This message is sent from the server to the client to store a bitmap in the bitmap cache

```
| message type (29) | cache_id uint32 | cache_index uint32 | data_length uint32 | data []byte |
```

> #### 30 - bitmap cache load

Reviewer: ```suggestion
##### 30 - bitmap cache load
```

This message is sent from the server to the client to render a bitmap stored in the bitmap cache

```
| message type (30) | cache_id uint32 | cache_index uint32 | top uint32 | left uint32 | right uint32 | bottom uint32 |
```

> Update to the UI is also required. The browser memory and javascript's array of hash tables, where cache id is the index of the array, and the cache index is a key in the hash table, will be used to store bitmap data.

Reviewer 2: Did you consider alternatives? Is a 40MB array a better option than something like IndexedDB?

Reviewer: I don't understand how the 40MB calculation is being made for the V2 cache. 

Looking at [2.2.7.1.4.2 Revision 2 (TS_BITMAPCACHE_CAPABILITYSET_REV2)](https://learn.microsoft.com/pt-br/openspecs/windows_protocols/ms-rdpbcgr/a5b9b9a6-5f67-4089-a95d-009bc8e25bfc), I see the ability to add up to 5 caches (`BitmapCacheNCellInfo`s), each with different limits for the number of entries. Adding up all of the maximum limits I get

600 + 600 + 65536 + 4096 + 2048 = 72880 entries

Assuming each entry means an (up to) 64x64 pixel chunk of the screen, at 32bpp that gives us 

(72880 entries) * (64x64 pixels/entry) * 32 (bits/pixel) = 9552527360 bits

which is 1.19 GB.

Author: Yes, I know, but at the same time, in the same specification they say something different here: https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/2bf92588-42bd-4527-8b3e-b90c56e292d2

> The Revision 1 bitmap cache defaults to 1500 KB of memory cache for 8 bpp, 3000 KB for 16 bpp, 4500 KB for 24 bpp, and 6000 KB for 32 bpp. The Revision 2 bitmap cache defaults to 10 MB of persistent cache storage for 8 bpp, 20 MB for 16 bpp, 30 MB for 24 bpp, and 40 MB for 32 bpp. The actual layout of the bitmap caches is specified in the Bitmap Cache Capability Sets ([MS-RDPBCGR] section [2.2.7.1.4](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/49ab3cb4-d51e-4fb4-a5a4-1beccc3cce6d)).

Author: Just to add to the above, we can control the number of cache entries that we support by changing cache numbers in the capability that you've mentioned and the server will respect that numbers.

Reviewer: Ah, I see. I suppose it might take some experimentation to figure out which it actually adheres to. Probably not worth the effort considering that we're not moving forward with the caching.

    Teleport Proxy->>Web Browser: tdp: store bitmap
    Web Browser->>cache in web browser memory: store bitmap
    Windows Server->>Teleport Windows Desktop Service: rdp: load bitmap
    Teleport Windows Desktop Service->>Teleport Proxy: tdp: load bitmap
    Teleport Proxy->>Web Browser: tdp: load bitmap
    cache in web browser memory->>Web Browser: load bitmap
```

- server sends [message](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/3c76f8fb-fcaa-4ca6-9d2e-3f449bdbed40) to inform client to store bitmap in the cache
- windows destkop service translates rdp message into tdp message to store bitmap and sends it to the proxy
- proxy forwards message to web browser
- web browser stores bitmap in the in-memory cache
- server sends [message](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/84c2ec2f-f776-405b-9b48-6894a28b1b14) to inform client to load bitmap from the cache
- windows destkop service translates rdp message into tdp message to load bitmap and sends it to the proxy
- proxy forwards message to web browser
- web browser loads bitmap from the in-memory cache

#### session recordings

> Since we no longer send every bitmap over the wire and to keep session recordings to work we'll also need to keep the cache of the bitmaps at the proxy. When we encounter the `save bitmap message` we'll need to store that bitmap at the proxy and then whenver `load bitmap message` will be sent we'll need to load bitmap from the cache and generate appropriate event with the bitmap data. 

Reviewer 2: Is this necessary?

If we record the messages that save to and load from the cache, we can play back the recording in the same way, right?

Author: It's really not necessary. The other option works as well.When I was thinking about what needs to be changed to make session recording work with new cache messages it looked like it might be an easier thing to do. The more I think about it now, the option with updating the logic for session recording will be cleaner solution, as we won't need to store anything on the proxy level. We'll  only need to update the function for recordings. Anyway, thanks for suggestion! I'll update this section.

### Process bitmaps in Rust library

While interacting with the remote desktop using the RDP protocol, most of the protocol messages exchanged between the server and a client are related to rendering bitmaps. Messages almost always contain compressed data to reduce bandwidth usage and latency. Rendering compressed bitmaps on the screen requires uncompressing data first using decompress algorithm.
 It is the most common operation while using the protocol, so it must execute extermaly fast to provide good user experience.

> The best performance can be achived be moving the bitmap decompression procdure from the Go client to the faster compiled library written in the Rust language. It'll also allow to simplyfy the way we encode and decode PNG tdp messages.

Reviewer 2: I thought decompression was already done in Rust. Can you link to the Go code you're speaking about here?

Author: What I really should've done is to add annotation that this is already implemented. My intention was to include it in in the RFD so that if anyone else is looking at this RFD will see what was important for us when we were checking for performance issues in the Desktop Access. I think this was really important so just wanted to include it here.

Reviewer 2: Decompression has always been done in Rust. Do you mean PNG encoding?

Author: Yes! For some reason I mixed these two terms. I'll revisit the doc and make sure that I'm using the right terms 😄
